code/modules/modular_computers/hardware/gprs_card.dm

```diff
@@ -0,0 +1,126 @@
+	. = ..()
+	SSpackets.remove_object(src, gprs_frequency, RADIO_PDAMESSAGE)
+	radio_connection = null
+	packet_queue.Cut()
+
+/obj/item/computer_hardware/network_card/packetnet/diagnostics(mob/user)
+	..()
+	to_chat(user, "GPRS - General Packet Radio Service")
+	to_chat(user, "┗ TX/RX Band - [format_frequency(gprs_frequency)]")
+
+/obj/item/computer_hardware/network_card/packetnet/examine(mob/user)
+	. = ..()
+	if(get_dist(get_turf(src), get_turf(user)) >= 2)
+		. += span_info("You're too far away to read the board...")
+		return
+	. += span_info("The silkscreen reads...")
+	. += "\t[span_info("GPRS FREQ: [span_robot(format_frequency(gprs_frequency))]")]"
+	. += "\t[span_info("GPRS DMEI: [span_robot(hardware_id)]")]"
+
+/obj/item/computer_hardware/network_card/packetnet/receive_signal(datum/signal/signal)
+	if(!holder || !signal.data) //Basic checks
+		return
+	if(signal.transmission_method != TRANSMISSION_RADIO)
+		CRASH("[src] Received non-radio packet, transmission method ID [signal.transmission_method], Expected [TRANSMISSION_RADIO]")
+	var/list/signal_data = signal.data //medium velocity silver hedgehog
+	var/signal_d_addr = signal_data["d_addr"]
+	if(signal_d_addr == NET_ADDRESS_PING) //Ping.
+		var/datum/signal/outgoing = new(
+			src,
+			list(
+				"d_addr" = signal_data["s_addr"],
+				"command" = NET_COMMAND_PING_REPLY,
+				"netclass" = NETCLASS_GRPS_CARD,
+				"netaddr" = hardware_id
+			)
+		)
+		if(signal_data["pda_scan"] && istype(holder, /obj/item/modular_computer/tablet))
+			// If we're on an actual tablet, pass along the user info. No privacy here.
+			var/obj/item/modular_computer/tablet/tab_holder = holder
+			var/list/og_data = outgoing.data
+			og_data["reg_name"] = tab_holder.saved_identification
+			og_data["reg_job"] = tab_holder.saved_job
+		post_signal(outgoing)
+	//Either it's broadcast or directed to us.
+	if(isnull(signal_d_addr) || signal_d_addr == hardware_id)
+		//We don't really care what it is, just store it.
+		push_signal(signal)
+
+
+/obj/item/computer_hardware/network_card/packetnet/proc/post_signal(datum/signal/signal)
+	if(!radio_connection || !signal)
+		return FALSE // Something went wrong.
+	radio_connection.post_signal(signal, RADIO_PDAMESSAGE)
+	return TRUE //We at least tried.
+
+/// Take a signal out of the queue.
+/obj/item/computer_hardware/network_card/packetnet/proc/pop_signal()
+	if(!length(packet_queue))
+		return FALSE //Nothing left, chief.
+	var/datum/signal/popped = packet_queue[1]
+	packet_queue -= popped
+	return popped
+
+/// Push a signal onto the queue, Drop a packet if we're over the limit.
+/obj/item/computer_hardware/network_card/packetnet/proc/push_signal(datum/signal/signal)
+	PRIVATE_PROC(TRUE)
+	if(signal.has_magic_data & MAGIC_DATA_MUST_DISCARD)
+		return //We can't hold volatile signals.
+	if(length(packet_queue) == queue_max)
+		pop_signal() //Discard the first signal in the queue
+	packet_queue += signal
+	return
```

Comment:
queues and stacks are different, you don't push to a queue, you append. Pop is only acceptable terminology here because python does this too.